File: services/entity-resolution-service/src/fusion_base/models/match_rule.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, types::Json as SqlJson};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockingStrategyConfig {
    pub strategy_type: String,
    pub key_fields: Vec<String>,
    pub window_size: i32,
    pub bucket_count: i32,
}

impl Default for BlockingStrategyConfig {
    fn default() -> Self {
        Self {
            strategy_type: "key-based".to_string(),
            key_fields: vec![
                "email".to_string(),
                "phone".to_string(),
                "display_name".to_string(),
            ],
            window_size: 5,
            bucket_count: 24,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MatchCondition {
    pub field: String,
    pub comparator: String,
    pub weight: f32,
    pub threshold: f32,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MatchRule {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub status: String,
    pub entity_type: String,
    pub blocking_strategy: BlockingStrategyConfig,
    pub conditions: Vec<MatchCondition>,
    pub review_threshold: f32,
    pub auto_merge_threshold: f32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateMatchRuleRequest {
    pub name: String,
    pub description: Option<String>,
    pub status: Option<String>,
    pub entity_type: Option<String>,
    pub blocking_strategy: Option<BlockingStrategyConfig>,
    pub conditions: Vec<MatchCondition>,
    pub review_threshold: Option<f32>,
    pub auto_merge_threshold: Option<f32>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateMatchRuleRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub status: Option<String>,
    pub entity_type: Option<String>,
    pub blocking_strategy: Option<BlockingStrategyConfig>,
    pub conditions: Option<Vec<MatchCondition>>,
    pub review_threshold: Option<f32>,
    pub auto_merge_threshold: Option<f32>,
}

#[derive(Debug, Clone, FromRow)]
pub struct MatchRuleRow {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub status: String,
    pub entity_type: String,
    pub blocking_strategy: SqlJson<BlockingStrategyConfig>,
    pub conditions: SqlJson<Vec<MatchCondition>>,
    pub review_threshold: f32,
    pub auto_merge_threshold: f32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<MatchRuleRow> for MatchRule {
    fn from(value: MatchRuleRow) -> Self {
        Self {
            id: value.id,
            name: value.name,
            description: value.description,
            status: value.status,
            entity_type: value.entity_type,
            blocking_strategy: value.blocking_strategy.0,
            conditions: value.conditions.0,
            review_threshold: value.review_threshold,
            auto_merge_threshold: value.auto_merge_threshold,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

File: services/entity-resolution-service/internal/models/match_rule.go
// Package models holds the wire-format types for entity-resolution-service.
//
// All struct tags match the Rust serde encoding (snake_case fields,
// f32 → float32). The Rust source has no migrations directory; the
// schema lives in the Go-side embedded migrations now (since the Go
// port is canonical — Rust binary is `fn main(){}`).
package models

import (
	"time"

	"github.com/google/uuid"
)

// ListResponse is the {"data": [...]} envelope used by all list endpoints.
type ListResponse[T any] struct {
	Data []T `json:"data"`
}

// ErrorResponse is the {"error": "..."} envelope.
type ErrorResponse struct {
	Error string `json:"error"`
}

// FusionOverview is the canonical summary payload.
type FusionOverview struct {
	RuleCount               int64 `json:"rule_count"`
	ActiveJobCount          int64 `json:"active_job_count"`
	CompletedJobCount       int64 `json:"completed_job_count"`
	ClusterCount            int64 `json:"cluster_count"`
	PendingReviewCount      int64 `json:"pending_review_count"`
	GoldenRecordCount       int64 `json:"golden_record_count"`
	AutoMergedClusterCount  int64 `json:"auto_merged_cluster_count"`
}

// --- Match rule -----------------------------------------------------------

// BlockingStrategyConfig defaults: strategy_type="key-based",
// key_fields=["email","phone","display_name"], window_size=5, bucket_count=24.
type BlockingStrategyConfig struct {
	StrategyType string   `json:"strategy_type"`
	KeyFields    []string `json:"key_fields"`
	WindowSize   int32    `json:"window_size"`
	BucketCount  int32    `json:"bucket_count"`
}

// DefaultBlockingStrategy mirrors the Rust impl Default for BlockingStrategyConfig.
func DefaultBlockingStrategy() BlockingStrategyConfig {
	return BlockingStrategyConfig{
		StrategyType: "key-based",
		KeyFields:    []string{"email", "phone", "display_name"},
		WindowSize:   5,
		BucketCount:  24,
	}
}

type MatchCondition struct {
	Field      string  `json:"field"`
	Comparator string  `json:"comparator"`
	Weight     float32 `json:"weight"`
	Threshold  float32 `json:"threshold"`
	Required   bool    `json:"required"`
}

type MatchRule struct {
	ID                  uuid.UUID              `json:"id"`
	Name                string                 `json:"name"`
	Description         string                 `json:"description"`
	Status              string                 `json:"status"`
	EntityType          string                 `json:"entity_type"`
	BlockingStrategy    BlockingStrategyConfig `json:"blocking_strategy"`
	Conditions          []MatchCondition       `json:"conditions"`
	ReviewThreshold     float32                `json:"review_threshold"`
	AutoMergeThreshold  float32                `json:"auto_merge_threshold"`
	CreatedAt           time.Time              `json:"created_at"`
	UpdatedAt           time.Time              `json:"updated_at"`
}

type CreateMatchRuleRequest struct {
	Name               string                  `json:"name"`
	Description        *string                 `json:"description,omitempty"`
	Status             *string                 `json:"status,omitempty"`
	EntityType         *string                 `json:"entity_type,omitempty"`
	BlockingStrategy   *BlockingStrategyConfig `json:"blocking_strategy,omitempty"`
	Conditions         []MatchCondition        `json:"conditions"`
	ReviewThreshold    *float32                `json:"review_threshold,omitempty"`
	AutoMergeThreshold *float32                `json:"auto_merge_threshold,omitempty"`
}

type UpdateMatchRuleRequest struct {
	Name               *string                 `json:"name,omitempty"`
	Description        *string                 `json:"description,omitempty"`
	Status             *string                 `json:"status,omitempty"`
	EntityType         *string                 `json:"entity_type,omitempty"`
	BlockingStrategy   *BlockingStrategyConfig `json:"blocking_strategy,omitempty"`
	Conditions         *[]MatchCondition       `json:"conditions,omitempty"`
	ReviewThreshold    *float32                `json:"review_threshold,omitempty"`
	AutoMergeThreshold *float32                `json:"auto_merge_threshold,omitempty"`
}

// --- Merge strategy -------------------------------------------------------

type SurvivorshipRule struct {
	Field          string   `json:"field"`
	Strategy       string   `json:"strategy"`
	SourcePriority []string `json:"source_priority"`
	Fallback       string   `json:"fallback"`
}

type MergeStrategy struct {
	ID              uuid.UUID          `json:"id"`
	Name            string             `json:"name"`
	Description     string             `json:"description"`
	Status          string             `json:"status"`
	EntityType      string             `json:"entity_type"`
	DefaultStrategy string             `json:"default_strategy"`
	Rules           []SurvivorshipRule `json:"rules"`
	CreatedAt       time.Time          `json:"created_at"`
	UpdatedAt       time.Time          `json:"updated_at"`
}

type CreateMergeStrategyRequest struct {
	Name            string             `json:"name"`
	Description     *string            `json:"description,omitempty"`
	Status          *string            `json:"status,omitempty"`
	EntityType      *string            `json:"entity_type,omitempty"`
	DefaultStrategy *string            `json:"default_strategy,omitempty"`
	Rules           []SurvivorshipRule `json:"rules"`
}

type UpdateMergeStrategyRequest struct {
	Name            *string             `json:"name,omitempty"`
	Description     *string             `json:"description,omitempty"`
	Status          *string             `json:"status,omitempty"`
	EntityType      *string             `json:"entity_type,omitempty"`
	DefaultStrategy *string             `json:"default_strategy,omitempty"`
	Rules           *[]SurvivorshipRule `json:"rules,omitempty"`
}
